# Capstone Project - Opening More Fitness Centers in Los Angeles, California

### IBM Applied Data Science Capstone by IBM/Coursera

# Table of Contents

+ Introduction: Business Problem
+ Data
+ Methodology
    + Data Gathering
    + Geocoding
    + Data Visualoization: Mapping Los Angeles
    + Exploring Neighborhoods With Foursquare API
    + Clustering Neighborhoods by Gym & Fitness Center Data
+ Results
+ Discussion
+ Conclusion

# Introduction: Business Problem

The goal of this project is to determine optimal locations for a new gym or fitness center within Los Angeles, California. Specifically, the end result is to compile a list of neighborhoods that can serve as target consumer areas for stakeholders interested in opening a gym or fitness center in Los Angeles, California. The neighborhoods of interest will be those with no gyms or fitness centers within 3000 meters from their center points. 

# Data

The following data resources will be used for this project:


+ The official list of neighborhoods in Los Angeles. 
    + Description: Names of each Los Angeles neighborhood are listed. The scope of this project does not extend to the greater Los Angeles area and is confined to the official neighborhood boundaries under Los Angeles. 
    + Data Source: geohub.lacity.org
+ Coordinates for each Los Angeles neighborhood.
    + Description: Latitude and longitude coordinates for the center of each Los Angeles neighborhood. These coordinates will be used to plot circle markers over a map of Los Angeles and to explore each of them.
    + Source: geopy.geocoder package
+ Venue Data in proximity of each Los Angeles neighborhood.
    + Description: Using Foursquare API a search query will be done against the coordinates for each Los Angeles neighborhood and venue data that is within a radius of 3000 meters of each neighborhood will be generated. This venue data will then be filtered to account for only data related to gyms and fitness centers.
    + Data source: Foursquare API

# Methodology

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!pip install geopy
!pip install geocoder
import geocoder
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 

from bs4 import BeautifulSoup # library to parse HTML and XML documents

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!pip install folium
import folium # map rendering library

print('Libraries imported.')

     |████████████████████████████████| 98 kB 7.3 MB/s  eta 0:00:01
     |████████████████████████████████| 93 kB 2.7 MB/s  eta 0:00:01
Libraries imported.


### Data Gathering

In [2]:
#link to LA neighborhood data uploaded in github repository
url = "https://github.com/c2barreto/Coursera_Captstone/blob/main/LA_Times_Neighborhood_Boundaries.csv" 

In [3]:
df_Neighbs = pd.read_html(url)

In [4]:
df_Neighbs[0]

Unnamed: 0  OBJECTID                    name
0           NaN         1         Adams-Normandie
1           NaN         2                  Arleta
2           NaN         3       Arlington Heights
3           NaN         4         Atwater Village
4           NaN         5  Baldwin Hills/Crenshaw
5           NaN         6                 Bel-Air
6           NaN         7           Beverly Crest
7           NaN         8           Beverly Grove
8           NaN         9             Beverlywood
9           NaN        10           Boyle Heights
10          NaN        11               Brentwood
11          NaN        12     Broadway-Manchester
12          NaN        13             Canoga Park
13          NaN        14                 Carthay
14          NaN        15         Central-Alameda
15          NaN        16            Century City
16          NaN        17              Chatsworth
17          NaN        18    Chatsworth Reservoir
18          NaN        19     Chesterfield Square
19          NaN        20           Cheviot Hills
20          NaN        21               Chinatown
21          NaN        22            Cypress Park
22          NaN        23                 Del Rey
23          NaN        24                Downtown
24          NaN        25              Eagle Rock
25          NaN        26          East Hollywood
26          NaN        27               Echo Park
27          NaN        28               El Sereno
28          NaN        29            Elysian Park
29          NaN        30          Elysian Valley
30          NaN        31                  Encino
31          NaN        32         Exposition Park
32          NaN        33                 Fairfax
33          NaN        34                Florence
34          NaN        35           Glassell Park
35          NaN        36           Gramercy Park
36          NaN        37           Granada Hills
37          NaN        38           Green Meadows
38          NaN        39           Griffith Park
39          NaN        40            Hancock Park
40          NaN        41              Hansen Dam
41          NaN        42             Harbor City
42          NaN        43          Harbor Gateway
43          NaN        44         Harvard Heights
44          NaN        45            Harvard Park
45          NaN        46           Highland Park
46          NaN        47  Historic South-Central
47          NaN        48         Hollywood Hills
48          NaN        49    Hollywood Hills West
49          NaN        50               Hollywood
50          NaN        51               Hyde Park
51          NaN        52          Jefferson Park
52          NaN        53               Koreatown
53          NaN        54       Lake View Terrace
54          NaN        55               Larchmont
55          NaN        56            Leimert Park
56          NaN        57         Lincoln Heights
57          NaN        58               Los Feliz
58          NaN        59       Manchester Square
59          NaN        60               Mar Vista
60          NaN        61                Mid-City
61          NaN        62            Mid-Wilshire
62          NaN        63           Mission Hills
63          NaN        64       Montecito Heights
64          NaN        65        Mount Washington
65          NaN        66             North Hills
66          NaN        67         North Hollywood
67          NaN        68              Northridge
68          NaN        69       Pacific Palisades
69          NaN        70                 Pacoima
70          NaN        71                   Palms
71          NaN        72           Panorama City
72          NaN        73          Pico-Robertson
73          NaN        74              Pico-Union
74          NaN        75           Playa del Rey
75          NaN        76             Playa Vista
76          NaN        77            Porter Ranch
77          NaN        78             Rancho Park
78          NaN        79                  Reseda
79   

In [5]:
#Pass table into a Pandas data frame.
df_LA = df_Neighbs[0]

In [6]:
#Drop unnecessary columns
df_LA.drop(["Unnamed: 0", "OBJECTID"], inplace=True, axis=1) 

In [7]:
#Rename column
df_LA.rename(columns = {'name':'Neighborhood'}, inplace = True)

In [8]:
df_LA.head()

Neighborhood
0         Adams-Normandie
1                  Arleta
2       Arlington Heights
3         Atwater Village
4  Baldwin Hills/Crenshaw

In [9]:
df_LA.shape

(114, 1)

### Geocoding

In [10]:
# defining a function to retrieve each LA neighborhood's coordinates
def get_latlng(LAneighborhood):
    # initialize a variable to None
    lat_lng_coords = None
    # loop until all neighborhood coordinates are retrieved using geocoder package
    while(lat_lng_coords is None):
        g = geocoder.arcgis('{}, Los Angeles, California'.format(LAneighborhood))
        lat_lng_coords = g.latlng
    return lat_lng_coords

In [11]:
# call the defined function to get the neighborhood coordinates and then store the values into a list
coordList = [ get_latlng(LAneighborhood) for LAneighborhood in df_LA["Neighborhood"].tolist() ]

In [12]:
coordList

[[34.07809005053511, -118.30119989733268],
 [34.24905000000007, -118.43348999999995],
 [34.03989000000007, -118.32541999999995],
 [34.11970000000008, -118.25886999999994],
 [34.01043000000004, -118.33665999999994],
 [34.08361000000008, -118.43482999999998],
 [34.09653268819107, -118.40329514585207],
 [34.09544836979449, -118.42608686997163],
 [34.05138000000005, -118.40132999999997],
 [34.04004000000003, -118.21049999999997],
 [34.06626000000006, -118.47032999999999],
 [33.960930000000076, -118.27881999999994],
 [34.20239000000004, -118.60155999999995],
 [34.17639681534724, -118.22274327896007],
 [33.997420000000034, -118.24777999999998],
 [34.05944000000005, -118.41280999999998],
 [34.257250000000056, -118.59100999999998],
 [34.241840000000025, -118.62292999999994],
 [33.99251000000004, -118.31284999999997],
 [34.03403000000003, -118.41055999999998],
 [34.062290000000075, -118.23836999999997],
 [34.09448000000003, -118.22677999999996],
 [33.99082000000004, -118.42061999999999],
 [34.0

In [13]:
# create temporary dataframe to populate the coordinates
df_temp = pd.DataFrame(coordList, columns=['Latitude', 'Longitude'])

In [14]:
# merge the temp dataframe with the df_LA
df_LA['Latitude'] = df_temp['Latitude']
df_LA['Longitude'] = df_temp['Longitude']

In [15]:
df_LA

Neighborhood   Latitude   Longitude
0           Adams-Normandie  34.078090 -118.301200
1                    Arleta  34.249050 -118.433490
2         Arlington Heights  34.039890 -118.325420
3           Atwater Village  34.119700 -118.258870
4    Baldwin Hills/Crenshaw  34.010430 -118.336660
5                   Bel-Air  34.083610 -118.434830
6             Beverly Crest  34.096533 -118.403295
7             Beverly Grove  34.095448 -118.426087
8               Beverlywood  34.051380 -118.401330
9             Boyle Heights  34.040040 -118.210500
10                Brentwood  34.066260 -118.470330
11      Broadway-Manchester  33.960930 -118.278820
12              Canoga Park  34.202390 -118.601560
13                  Carthay  34.176397 -118.222743
14          Central-Alameda  33.997420 -118.247780
15             Century City  34.059440 -118.412810
16               Chatsworth  34.257250 -118.591010
17     Chatsworth Reservoir  34.241840 -118.622930
18      Chesterfield Square  33.992510 -118.312850
19            Cheviot Hills  34.034030 -118.410560
20                Chinatown  34.062290 -118.238370
21             Cypress Park  34.094480 -118.226780
22                  Del Rey  33.990820 -118.420620
23                 Downtown  34.048340 -118.255640
24               Eagle Rock  34.139270 -118.210870
25           East Hollywood  34.090860 -118.293350
26                Echo Park  34.076090 -118.255810
27                El Sereno  34.076850 -118.179340
28             Elysian Park  34.082500 -118.233200
29           Elysian Valley  34.071738 -117.721657
30                   Encino  34.165380 -118.527110
31          Exposition Park  34.011410 -118.289970
32                  Fairfax  34.076100 -118.361220
33                 Florence  33.974750 -118.245390
34            Glassell Park  34.119220 -118.230440
35            Gramercy Park  34.033900 -118.312580
36            Granada Hills  34.263940 -118.497940
37            Green Meadows  33.812272 -118.308367
38            Griffith Park  34.123270 -118.299900
39             Hancock Park  34.068920 -118.332550
40               Hansen Dam  34.260560 -118.385360
41              Harbor City  33.797810 -118.302480
42           Harbor Gateway  33.854184 -118.303480
43          Harvard Heights  34.043390 -118.309030
44             Harvard Park  34.144400 -118.252970
45            Highland Park  34.109820 -118.192320
46   Historic South-Central  34.019790 -118.257280
47          Hollywood Hills  34.105240 -118.320650
48     Hollywood Hills West  34.120060 -118.350900
49                Hollywood  34.101560 -118.337080
50                Hyde Park  33.989040 -118.326700
51           Jefferson Park  34.032640 -118.318020
52                Koreatown  34.057790 -118.300910
53        Lake View Terrace  34.274530 -118.373500
54                Larchmont  34.080990 -118.323650
55             Leimert Park  34.008380 -118.330450
56          Lincoln Heights  34.069830 -118.211640
57                Los Feliz  34.105410 -118.289240
58        Manchester Square  34.563475 -118.050453
59                Mar Vista  34.004680 -118.429130
60                 Mid-City  34.040170 -118.347630
61             Mid-Wilshire  34.052863 -118.458251
62            Mission Hills  34.273970 -118.467830
63        Montecito Heights  34.091980 -118.201010
64         Mount Washington  34.099040 -118.211340
65              North Hills  34.235630 -118.484700
66          North Hollywood  34.169820 -118.378990
67               Northridge  34.235040 -118.533860
68        Pacific Palisades  34.047100 -118.526880
69                  Pacoima  34.261080 -118.428980
70                    Palms  34.019860 -118.420960
71            Panorama City  34.223020 -118.448750
72           Pico-Robertson  34.055750 -118.380000
73               Pico-Union  34.040040 -118.285550
74            Playa del Rey  33.959240 -118.448010
75              Playa Vista  33.973950 -118.430380
76             Porter Ranch  34.280270 -118.561300
77              Rancho Park  34.04507

In [16]:
df_LA.to_csv("df_LA3.csv", index=False)

### Data Visualization: Mapping Los Angeles

In [17]:
# Get the coordinates of Los Angeles
address = 'Los Angeles, CA'

geolocator = Nominatim(user_agent="ca_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Los Angeles are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Los Angeles are 34.0536909, -118.2427666.


In [18]:
# create map of Toronto using latitude and longitude values
map_LA = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, neighborhood in zip(df_LA['Latitude'], df_LA['Longitude'], df_LA['Neighborhood']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_LA)  
    
map_LA

In [19]:
# save the map as HTML file
map_LA.save('map_LA3.html')

### Exploring Neighborhoods With Foursquare API

In [20]:
# defining foursquare credentials and version
CLIENT_ID = 'H0ZIZZAEGBFDEGUJYC0XZKX4Z2ROHFX53WN0V5SENBVMDBVU' # my Foursquare ID
CLIENT_SECRET = 'D1HOVQ0X0UQV1JHXSZ0TIC0HFODDSELZH2WTWLJ0U5GQ5E1A' # my Foursquare Secret
VERSION = '20201105' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: H0ZIZZAEGBFDEGUJYC0XZKX4Z2ROHFX53WN0V5SENBVMDBVU
CLIENT_SECRET:D1HOVQ0X0UQV1JHXSZ0TIC0HFODDSELZH2WTWLJ0U5GQ5E1A


In [21]:
#defining a function to get top venues of each neighborhood within a 500 meter radius
def getNearbyVenues(names, latitudes, longitudes, radius=2000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
        
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [22]:
# code to run the above function on each neighborhood and create a new dataframe called losangeles_venues
losangeles_venues = getNearbyVenues(names=df_LA['Neighborhood'],
                                   latitudes=df_LA['Latitude'],
                                   longitudes=df_LA['Longitude']
                                  )

Adams-Normandie
Arleta
Arlington Heights
Atwater Village
Baldwin Hills/Crenshaw
Bel-Air
Beverly Crest
Beverly Grove
Beverlywood
Boyle Heights
Brentwood
Broadway-Manchester
Canoga Park
Carthay
Central-Alameda
Century City
Chatsworth
Chatsworth Reservoir
Chesterfield Square
Cheviot Hills
Chinatown
Cypress Park
Del Rey
Downtown
Eagle Rock
East Hollywood
Echo Park
El Sereno
Elysian Park
Elysian Valley
Encino
Exposition Park
Fairfax
Florence
Glassell Park
Gramercy Park
Granada Hills
Green Meadows
Griffith Park
Hancock Park
Hansen Dam
Harbor City
Harbor Gateway
Harvard Heights
Harvard Park
Highland Park
Historic South-Central
Hollywood Hills
Hollywood Hills West
Hollywood
Hyde Park
Jefferson Park
Koreatown
Lake View Terrace
Larchmont
Leimert Park
Lincoln Heights
Los Feliz
Manchester Square
Mar Vista
Mid-City
Mid-Wilshire
Mission Hills
Montecito Heights
Mount Washington
North Hills
North Hollywood
Northridge
Pacific Palisades
Pacoima
Palms
Panorama City
Pico-Robertson
Pico-Union
Playa del Rey

In [23]:
print(losangeles_venues.shape)
losangeles_venues.head()

(9586, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Adams-Normandie               34.07809               -118.3012   
1  Adams-Normandie               34.07809               -118.3012   
2  Adams-Normandie               34.07809               -118.3012   
3  Adams-Normandie               34.07809               -118.3012   
4  Adams-Normandie               34.07809               -118.3012   

                                               Venue  Venue Latitude  \
0                                        Noshi Sushi       34.076159   
1                                            Jaraguá       34.076364   
2  Kim Sun Young Hair Beauty Salon (Kim Sun Young...       34.076453   
3                                Guatemalteca Bakery       34.076303   
4                                Cactus Mexican Food       34.076194   

   Venue Longitude      Venue Category  
0      -118.305374    Sushi Restaurant  
1      -118.306646        Cocktail Bar  
2      -118.308921  Salon / Barbershop  
3      -118.297168          Restaurant  
4      -118.304147  Mexican Restaurant

In [24]:
#Checking how many venues were returned for each neighborhood
losangeles_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                                   
Adams-Normandie                           100                     100    100   
Arleta                                     56                      56     56   
Arlington Heights                         100                     100    100   
Atwater Village                           100                     100    100   
Baldwin Hills/Crenshaw                     75                      75     75   
Bel-Air                                    62                      62     62   
Beverly Crest                              41                      41     41   
Beverly Grove                              19                      19     19   
Beverlywood                               100                     100    100   
Boyle Heights                             100                     100    100   
Brentwood                                  94                      94     94   
Broadway-Manchester                        53                      53     53   
Canoga Park                               100                     100    100   
Carthay                                    21                      21     21   
Central-Alameda                            48                      48     48   
Century City                              100                     100    100   
Chatsworth                                 99                      99     99   
Chatsworth Reservoir                       23                      23     23   
Chesterfield Square                        62                      62     62   
Cheviot Hills                             100                     100    100   
Chinatown                                 100                     100    100   
Cypress Park                               60                      60     60   
Del Rey                                   100                     100    100   
Downtown                                  100                     100    100   
Eagle Rock                                100                     100    100   
East Hollywood                            100                     100    100   
Echo Park                                 100                     100    100   
El Sereno                                  58                      58     58   
Elysian Park                              100                     100    100   
Elysian Valley                             54                      54     54   
Encino                                    100                     100    100   
Exposition Park                           100                     100    100   
Fairfax                                   100                     100    100   
Florence                                   98                      98     98   
Glassell Park                              78                      78     78   
Gramercy Park                              94                      94     94   
Granada Hills                             100                     100    100   
Green Meadows                              96                      96     96   
Griffith Park                              75                      75     75   
Hancock Park                              100                     100    100   
Hansen Dam                                 22                      22     22   
Harbor City                                93                      93     93   
Harbor Gateway                             92                      92     92   
Harvard Heights                           100                     100    100   
Harvard Park                              100                     100    100   
Highland Park                             100                     100    100   
Historic South-Central                     88                      88     88   
Hollywood                                 100                     100    100   
Hollywood Hills         

In [25]:
# checking how many unique categories can be curated from all the returned venues
print('There are {} uniques categories.'.format(len(losangeles_venues['Venue Category'].unique())))

There are 413 uniques categories.


In [27]:
# one hot encoding
losangeles_onehot = pd.get_dummies(losangeles_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
losangeles_onehot['Neighborhood'] = losangeles_venues['Neighborhood'] 

# move neighborhood column to the first column
# define a list of column names
cols = losangeles_onehot.columns.tolist()
cols

# move the column name to the beggining
cols.insert(0, cols.pop(cols.index('Neighborhood')))
cols

#then use .reindex() function to reorder
losangeles_onehot = losangeles_onehot.reindex(columns= cols)

#check result
losangeles_onehot.head()


Neighborhood  ATM  Accessories Store  Adult Boutique  Airport  \
0  Adams-Normandie    0                  0               0        0   
1  Adams-Normandie    0                  0               0        0   
2  Adams-Normandie    0                  0               0        0   
3  Adams-Normandie    0                  0               0        0   
4  Adams-Normandie    0                  0               0        0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Amphitheater  Andhra Restaurant  Antique Shop  Aquarium  Arcade  \
0             0                  0             0         0       0   
1             0                  0             0         0       0   
2             0                  0             0         0       0   
3             0                  0             0         0       0   
4             0                  0             0         0       0   

   Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                       0            0           0                    0   
1                       0            0           0                    0   
2                       0            0           0                    0   
3                       0            0           0                    0   
4                       0            0           0                    0   

   Asian Restaurant  Astrologer  Athletics & Sports  Auditorium  \
0                 0           0                   0           0   
1                 0           0                   0           0   
2                 0           0                   0           0   
3                 0           0                   0           0   
4                 0           0                   0           0   

   Australian Restaurant  Auto Dealership  Auto Garage  Automotive Shop  \
0                      0                0            0                0   
1                      0                0            0                0   
2                      0                0            0                0   
3                      0                0            0                0   
4                      0                0            0                0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  \
0                 0                 0                   0      0   
1                 0                 0                   0      0   
2                 0                 0                   0      0   
3                 0                 0                   0      0   
4                 0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  Big Box Store  \
0                0         0            0           0              0   
1                0         0            0           0              0   
2                0         0            0           0              0   
3                0         0            0           0              0   
4                0         0            0           0              0   

   Bike Rental / Bike Share  Bike Shop  Bike Trail  Bistro  Board Shop  \
0                      

In [28]:
#viewing the new dataframe size
losangeles_onehot.shape

(9586, 413)

In [29]:
#Grouping the rows by neighborhood and by taking the mean of the frequency of occurrence of each category
losangeles_grouped = losangeles_onehot.groupby('Neighborhood').mean().reset_index()
losangeles_grouped

Neighborhood       ATM  Accessories Store  Adult Boutique  \
0           Adams-Normandie  0.000000               0.00            0.00   
1                    Arleta  0.000000               0.00            0.00   
2         Arlington Heights  0.000000               0.00            0.00   
3           Atwater Village  0.010000               0.00            0.00   
4    Baldwin Hills/Crenshaw  0.000000               0.00            0.00   
5                   Bel-Air  0.000000               0.00            0.00   
6             Beverly Crest  0.000000               0.00            0.00   
7             Beverly Grove  0.000000               0.00            0.00   
8               Beverlywood  0.000000               0.00            0.00   
9             Boyle Heights  0.010000               0.00            0.00   
10                Brentwood  0.000000               0.00            0.00   
11      Broadway-Manchester  0.000000               0.00            0.00   
12              Canoga Park  0.000000               0.00            0.00   
13                  Carthay  0.000000               0.00            0.00   
14          Central-Alameda  0.000000               0.00            0.00   
15             Century City  0.000000               0.00            0.00   
16               Chatsworth  0.000000               0.00            0.00   
17     Chatsworth Reservoir  0.000000               0.00            0.00   
18      Chesterfield Square  0.016129               0.00            0.00   
19            Cheviot Hills  0.000000               0.00            0.00   
20                Chinatown  0.000000               0.00            0.00   
21             Cypress Park  0.000000               0.00            0.00   
22                  Del Rey  0.000000               0.00            0.00   
23                 Downtown  0.000000               0.00            0.00   
24               Eagle Rock  0.000000               0.00            0.00   
25           East Hollywood  0.000000               0.00            0.01   
26                Echo Park  0.000000               0.00            0.00   
27                El Sereno  0.000000               0.00            0.00   
28             Elysian Park  0.000000               0.00            0.00   
29           Elysian Valley  0.018519               0.00            0.00   
30                   Encino  0.000000               0.00            0.00   
31          Exposition Park  0.000000               0.00            0.00   
32                  Fairfax  0.000000               0.00            0.00   
33                 Florence  0.000000               0.00            0.00   
34            Glassell Park  0.012821               0.00            0.00   
35            Gramercy Park  0.010638               0.00            0.00   
36            Granada Hills  0.000000               0.00            0.00   
37            Green Meadows  0.010417               0.00            0.00   
38            Griffith Park  0.000000               0.00            0.00   
39             Hancock Park  0.000000               0.01            0.00   
40               Hansen Dam  0.000000               0.00            0.00   
41              Harbor City  0.000000               0.00            0.00   
42           Harbor Gateway  0.000000               0.00            0.00   
43          Harvard Heights  0.000000               0.00            0.00   
44             Harvard Park  0.000000               0.00            0.00   
45            Highland Park  0.000000               0.00            0.00   
46   Historic South-Central  0.000000               0.00            0.00   
47                Hollywood  0.000000               0.00            0.00   
48          Hollywood Hills  0.000000               0.00            0.00   
49     Hollywood Hills West  0.000000               0.00            0.00   
50                Hyde Park  0.010870               0.00            0.00   
51           Jefferson Park  0.000000               0.00       

### Clustering Neighborhoods by Gym & Fitness Center Data

In [30]:
#confirming new size
losangeles_grouped.shape

(114, 413)

In [31]:
# count of neighborhoods with at least 1 occurance of a gym or fitness center near them
len(losangeles_grouped[losangeles_grouped["Gym / Fitness Center"] > 0])

47

In [32]:
# new data frame for LA neighborhood data on Gyms and Fitness Centers
LA_gyms = losangeles_grouped[["Neighborhood","Gym / Fitness Center"]]

In [33]:
LA_gyms.head()

Neighborhood  Gym / Fitness Center
0         Adams-Normandie                  0.00
1                  Arleta                  0.00
2       Arlington Heights                  0.00
3         Atwater Village                  0.02
4  Baldwin Hills/Crenshaw                  0.00

In [34]:
# set number of clusters
kclusters = 5

losangeles_grouped_clustering = LA_gyms.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(losangeles_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:100] 

array([1, 1, 1, 3, 1, 1, 3, 1, 1, 1, 2, 1, 1, 4, 1, 1, 0, 1, 3, 3, 1, 1,
       2, 0, 0, 1, 3, 1, 1, 1, 0, 1, 0, 1, 4, 1, 1, 1, 1, 0, 1, 0, 1, 1,
       0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1,
       0, 3, 1, 1, 1, 3, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 4, 3, 1, 2,
       1, 1, 1, 3, 1, 1, 1, 0, 0, 1, 2, 1], dtype=int32)

In [35]:
LA_merged = LA_gyms.copy()
# add clustering labels column 
LA_merged["Cluster Labels"] = kmeans.labels_

LA_merged.head() # check the last columns!

Neighborhood  Gym / Fitness Center  Cluster Labels
0         Adams-Normandie                  0.00               1
1                  Arleta                  0.00               1
2       Arlington Heights                  0.00               1
3         Atwater Village                  0.02               3
4  Baldwin Hills/Crenshaw                  0.00               1

In [36]:
# merge df_LA with LA_merged to add latitude/longitude for each neighborhood
LA_merged = LA_merged.join(df_LA.set_index("Neighborhood"), on="Neighborhood")

print(LA_merged.shape)
LA_merged.head() # check the last columns!

(114, 5)


Neighborhood  Gym / Fitness Center  Cluster Labels  Latitude  \
0         Adams-Normandie                  0.00               1  34.07809   
1                  Arleta                  0.00               1  34.24905   
2       Arlington Heights                  0.00               1  34.03989   
3         Atwater Village                  0.02               3  34.11970   
4  Baldwin Hills/Crenshaw                  0.00               1  34.01043   

   Longitude  
0 -118.30120  
1 -118.43349  
2 -118.32542  
3 -118.25887  
4 -118.33666

In [37]:
#visualize the resulting clusters
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(LA_merged['Latitude'], LA_merged['Longitude'], LA_merged['Neighborhood'], LA_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [38]:
map_clusters.save('map_clusters.html')

# Results

### Cluster 0 (Red)

In [41]:
cluster0 = LA_merged.loc[LA_merged['Cluster Labels'] == 0]
cluster0

Neighborhood  Gym / Fitness Center  Cluster Labels   Latitude  \
16         Chatsworth              0.010101               0  34.257250   
23           Downtown              0.010000               0  34.048340   
24         Eagle Rock              0.010000               0  34.139270   
30             Encino              0.010000               0  34.165380   
32            Fairfax              0.010000               0  34.076100   
39       Hancock Park              0.010000               0  34.068920   
41        Harbor City              0.010753               0  33.797810   
44       Harvard Park              0.010000               0  34.144400   
45      Highland Park              0.010000               0  34.109820   
47          Hollywood              0.010000               0  34.101560   
48    Hollywood Hills              0.010000               0  34.105240   
60          Mar Vista              0.010000               0  34.004680   
61           Mid-City              0.010000               0  34.040170   
62       Mid-Wilshire              0.010000               0  34.052863   
63      Mission Hills              0.014286               0  34.273970   
66        North Hills              0.011765               0  34.235630   
73     Pico-Robertson              0.010000               0  34.055750   
75        Playa Vista              0.010000               0  33.973950   
78        Rancho Park              0.010000               0  34.045070   
80          San Pedro              0.011494               0  33.737870   
81           Sawtelle              0.010000               0  34.030570   
82    Sepulveda Basin              0.010000               0  34.023755   
95        Valley Glen              0.010000               0  34.187110   
96     Valley Village              0.012658               0  34.164880   
104        West Adams              0.010000               0  34.027980   
105        West Hills              0.010000               0  34.200360   
106  West Los Angeles              0.010000               0  34.046470   
107       Westchester              0.010000               0  33.959730   
108          Westlake              0.010000               0  34.053710   
112          Winnetka              0.012048               0  34.203290   

      Longitude  
16  -118.591010  
23  -118.255640  
24  -118.210870  
30  -118.527110  
32  -118.361220  
39  -118.332550  
41  -118.302480  
44  -118.252970  
45  -118.192320  
47  -118.337080  
48  -118.320650  
60  -118.429130  
61  -118.347630  
62  -118.458251  
63  -118.467830  
66  -118.484700  
73  -118.380000  
75  -118.430380  
78  -118.419240  
80  -118.284610  
81  -118.463410  
82  -118.425995  
95  -118.413810  
96  -118.398030  
104 -118.353140  
105 -118.629330  
106 -118.433950  
107 -118.413690  
108 -118.271300  
112 -118.570980

In [42]:
cluster0.shape

(30, 5)

Cluster 0 groups together 30 neighborhoods with a low concentration of fitness centers in proximity. The venue data for these neighborhoods generated at least 1 occurrence of a fitness center within a 2000 meter radius from their center points.

### Cluster 1 (Purple)

In [43]:
cluster1 = LA_merged.loc[LA_merged['Cluster Labels'] == 1]
cluster1

Neighborhood  Gym / Fitness Center  Cluster Labels   Latitude  \
0           Adams-Normandie                   0.0               1  34.078090   
1                    Arleta                   0.0               1  34.249050   
2         Arlington Heights                   0.0               1  34.039890   
4    Baldwin Hills/Crenshaw                   0.0               1  34.010430   
5                   Bel-Air                   0.0               1  34.083610   
7             Beverly Grove                   0.0               1  34.095448   
8               Beverlywood                   0.0               1  34.051380   
9             Boyle Heights                   0.0               1  34.040040   
11      Broadway-Manchester                   0.0               1  33.960930   
12              Canoga Park                   0.0               1  34.202390   
14          Central-Alameda                   0.0               1  33.997420   
15             Century City                   0.0               1  34.059440   
17     Chatsworth Reservoir                   0.0               1  34.241840   
20                Chinatown                   0.0               1  34.062290   
21             Cypress Park                   0.0               1  34.094480   
25           East Hollywood                   0.0               1  34.090860   
27                El Sereno                   0.0               1  34.076850   
28             Elysian Park                   0.0               1  34.082500   
29           Elysian Valley                   0.0               1  34.071738   
31          Exposition Park                   0.0               1  34.011410   
33                 Florence                   0.0               1  33.974750   
35            Gramercy Park                   0.0               1  34.033900   
36            Granada Hills                   0.0               1  34.263940   
37            Green Meadows                   0.0               1  33.812272   
38            Griffith Park                   0.0               1  34.123270   
40               Hansen Dam                   0.0               1  34.260560   
42           Harbor Gateway                   0.0               1  33.854184   
43          Harvard Heights                   0.0               1  34.043390   
46   Historic South-Central                   0.0               1  34.019790   
49     Hollywood Hills West                   0.0               1  34.120060   
50                Hyde Park                   0.0               1  33.989040   
51           Jefferson Park                   0.0               1  34.032640   
52                Koreatown                   0.0               1  34.057790   
53              Lake Balboa                   0.0               1  34.207160   
54        Lake View Terrace                   0.0               1  34.274530   
55                Larchmont                   0.0               1  34.080990   
56             Leimert Park                   0.0               1  34.008380   
57          Lincoln Heights                   0.0               1  34.069830   
58                Los Feliz                   0.0               1  34.105410   
59        Manchester Square                   0.0               1  34.563475   
64        Montecito Heights                   0.0               1  34.091980   
65         Mount Washington                   0.0               1  34.099040   
68               Northridge                   0.0               1  34.235040   
69        Pacific Palisades                   0.0               1  34.047100   
70                  Pacoima                   0.0               1  34.261080   
72            Panorama City                   0.0               1  34.223020   
74               Pico-Union                   0.0               1  34.040040   
76            Playa del Rey                   0.0               1  33.959240   
77             Porter Ranch                   0.0               1  34.280270   
79             

In [44]:
cluster1.shape

(67, 5)

Cluster 1 groups together 67 neighborhoods with no existence of fitness centers in proximity. The venue data for these neighborhoods generated 0 occurrences of a fitness center within a 2000 meter radius from their center points.

### Cluster 2 (Blue)

In [45]:
cluster2 = LA_merged.loc[LA_merged['Cluster Labels'] == 2]
cluster2

Neighborhood  Gym / Fitness Center  Cluster Labels  Latitude  Longitude
10    Brentwood              0.031915               2  34.06626 -118.47033
22      Del Rey              0.030000               2  33.99082 -118.42062
87  Studio City              0.030000               2  34.14453 -118.39561
98       Venice              0.030000               2  33.98754 -118.47215

In [46]:
cluster2.shape

(4, 5)

Cluster 2 groups together 4 neighborhoods with a mid-high concentration of fitness centers in proximity. The venue data for these neighborhoods generated 3 occurrences of a fitness center within a 2000 meter radius from their center points.

## Cluster 3 (Green)

In [47]:
cluster3 = LA_merged.loc[LA_merged['Cluster Labels'] == 3]
cluster3

Neighborhood  Gym / Fitness Center  Cluster Labels   Latitude  \
3        Atwater Village              0.020000               3  34.119700   
6          Beverly Crest              0.024390               3  34.096533   
18   Chesterfield Square              0.016129               3  33.992510   
19         Cheviot Hills              0.020000               3  34.034030   
26             Echo Park              0.020000               3  34.076090   
67       North Hollywood              0.020000               3  34.169820   
71                 Palms              0.020000               3  34.019860   
85           Silver Lake              0.020000               3  34.082730   
91               Tarzana              0.020000               3  34.175290   
113       Woodland Hills              0.020000               3  34.168890   

      Longitude  
3   -118.258870  
6   -118.403295  
18  -118.312850  
19  -118.410560  
26  -118.255810  
67  -118.378990  
71  -118.420960  
85  -118.273070  
91  -118.550100  
113 -118.611480

In [48]:
cluster3.shape

(10, 5)

Cluster 3 groups together 10 neighborhoods with a low-mid concentration of fitness centers in proximity. The venue data for these neighborhoods generated at least 2 occurrences of a fitness center within a 2000 meter radius from their center points.

### Cluster 4 (Orange)

In [49]:
cluster4 = LA_merged.loc[LA_merged['Cluster Labels'] == 4]
cluster4

Neighborhood  Gym / Fitness Center  Cluster Labels   Latitude   Longitude
13        Carthay              0.047619               4  34.176397 -118.222743
34  Glassell Park              0.038462               4  34.119220 -118.230440
84   Sherman Oaks              0.040000               4  34.151060 -118.444340

In [50]:
cluster4.shape

(3, 5)

Cluster 4 groups together 3 neighborhoods with a high concentration of fitness centers in proximity. The venue data for these neighborhoods generated at least 4 occurrences of a fitness center within a 2000 meter radius from their center points.

# Discussion

Based on the clustering data over the map of Los Angeles, the majority of neighborhoods under Los Angeles county lack access to nearby fitness centers. Very few neighborhoods harbor a mid to high concentration of local fitness centers present. In fact, clusters 2, 3 and 4, which contain a mid to high concentration of fitness centers, can be combined and they would only total 17 neighborhoods out of the 114. Unsurprisingly, clusters 2, 3, and 4 contain suburban neighborhoods with a higher income population. Judging also by their higher concentration of fitness centers, it is safe to recommend to avoid building new fitness centers in these areas since they already have competition established. Even more so, the intent is to give more local access to neighborhoods that lack fitness centers.

Continuing on, the second largest cluster is cluster 0, which list neighborhoods with at least one occurrence of a fitness center in proximity. Cluster 0 contained 30 neighborhoods out of the 114 in Los Angeles and geographically was concentrated towards Western Los Angeles and North West Los Angeles. Competition for memberships is probably not high in these neighborhoods so they can be considered as a secondary tier of neighborhoods for building more fitness centers around. 

Every neighborhood that did not generate an occurence of a fitness center within a 2000 meter radius from their center point was assigned to Cluster 1. Cluster 1 had a total of 67 out of 114 neighborhoods assigned to its cluster. Geographically, the neighborhoods under cluster 1 that are located in the most southern side of Los Angeles have the least availability to access a fitness center in proximity.This discrepancy could be due to the fact high concentration of low income residents reside in the southern side of Los Angeles which in the past might have deterred property developers from establishing fitness centers there. 

In Addition, there are also several other notable geographic concentrations from cluster 1 that display a lack of available neighborhoods such as East Los Angeles, Central Los Angeles, and the most northern regions of Los Angeles that hug San Fernando. These city regions from cluster 1 are also high population dense areas so there is definitely a high probability of demand for more convenient fitness centers in these areas. 



# Conclusion

Despite there being some limitations as far as how much venue data can be queried from Foursquare, enough venue data was gathered where an accurate mean frequency of occurrence for gyms and fitness centers can be obtained for each Los Angeles neighborhood. From these mean frequency of occurrences for gyms and fitness centers, five unique clusters were able to be made and each held notable geographic insights. In Particular the most stand out cluster that can be referred to as the ideal list of neighborhoods to target for building new fitness centers is cluster 1. Cluster 1 revealed that there are at least three big and populous regions in Los Angeles that have zero fitness centers established and they are Southern Los Angeles, East los Angeles, and Central Los Angeles. While these regions have historically housed low-income residents, this factor should not limit property developers from pursuing to build fitness centers in these regions. In fact, property developers should aim to build a market of low-cost membership fitness centers in these regions. Better yet, the city of Los Angeles can serve the health needs of its low-income residents better by building community fitness recreational centers in these neighborhoods as well. Overall, there is a market and need to build more fitness centers for the neighborhoods listed under cluster 1. 
